kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/PulsarStorageProducerIdManagerImpl.java

```diff
@@ -0,0 +1,210 @@
+    private final AtomicLong nextId = new AtomicLong(0);
+    private final String topic;
+    private final int blockSize;
+    private final PulsarClient pulsarClient;
+    private CompletableFuture<Reader<byte[]>> reader;
+    private CompletableFuture<Void> currentReadHandle;
+    private final ConcurrentLinkedQueue<Long> availableIdsLocally = new ConcurrentLinkedQueue<>();
+
+    private synchronized CompletableFuture<Reader<byte[]>> ensureReaderHandle() {
+        if (reader == null) {
+            reader = pulsarClient.newReader()
+                    .topic(topic)
+                    .startMessageId(MessageId.earliest)
+                    .startMessageIdInclusive()
+                    .readCompacted(true)
+                    .createAsync();
+        }
+        return reader;
+    }
+
+    private CompletableFuture<Void> readNextMessageIfAvailable(Reader<byte[]> reader) {
+        return reader
+                .hasMessageAvailableAsync()
+                .thenCompose(hasMessageAvailable -> {
+                    if (hasMessageAvailable == null
+                            || !hasMessageAvailable) {
+                        return CompletableFuture.completedFuture(null);
+                    } else {
+                        CompletableFuture<Message<byte[]>> opMessage = reader.readNextAsync();
+                        return opMessage.thenCompose(msg -> {
+                            byte[] value = msg.getValue();
+                            long newId = new BigInteger(value).longValue();
+                            log.debug("{} Read {} from {}", this, newId, msg.getMessageId());
+                            nextId.set(newId);
+                            return readNextMessageIfAvailable(reader);
+                        });
+                    }
+                });
+    }
+
+    private synchronized CompletableFuture<Void> ensureLatestData(boolean beforeWrite) {
+        if (currentReadHandle != null) {
+            if (beforeWrite) {
+                // we are inside a write loop, so
+                // we must ensure that we start to read now
+                // otherwise the write would use non up-to-date data
+                // so let's finish the current loop
+                log.debug("A read was already pending, starting a new one in order to ensure consistency");
+                return currentReadHandle
+                        .thenCompose(___ -> ensureLatestData(false));
+            }
+            // if there is an ongoing read operation then complete it
+            return currentReadHandle;
+        }
+        // please note that the read operation is async,
+        // and it is not execute inside this synchronized block
+        CompletableFuture<Reader<byte[]>> readerHandle = ensureReaderHandle();
+        final CompletableFuture<Void> newReadHandle =
+                readerHandle.thenCompose(this::readNextMessageIfAvailable);
+        currentReadHandle = newReadHandle;
```

Comment:
I see this method returns `currentReadHandle` when it's not null. Should here be following code?

```java
currentReadHandle = newReadHandle.thenApply(/* ... */);
return currentReadHandle;
```

Reply:
good idea.
I have updated the patch.
And I also now currentReadHandle will be the same CompletableFuture returned in case of non null `currentReadHandle`.

Reply:
I see when `currentReadHandle` is not null, the `ensureLatestData` returns `currentReadHandle`. Otherwise, it returns the `currentReadHandle.thenApply(/* ... */)`. Is it intentional?

I mean, the correct implementation might be

```java
        currentReadHandle = newReadHandle.thenApply((__) -> {
            endReadLoop(newReadHandle);
            return null;
        });
        return currentReadHandle;
```

Reply:
I tried to do what you are suggesting 10 minutes ago, but if I do saw than tests are not passing anymore :) 

the `endReadLoop` is only a cleanup operation.

- When you are "reading" then it doesn't matter which "currentReadHandle" you pick
- When you are "writing" then we must be sure that we pick the latest "currentReadHandle" and we complete the read, because we are inside the Exclusive Producer guard and nobody else (even in the same JVM) can write.

Reply:
Oh, sorry, I understand now.